<?xml version="1.0" encoding="UTF-8"?>
<launch>
	<arg name="map_frame_id" default="map" />
	<arg name="map_odom_only_frame_id" default="map_odom_only" />
	<arg name="odom_frame_id" default="odom" />
	<arg name="base_link_frame_id" default="base_footprint" />
	<arg name="localization_ground_truth_frame_id" default="map_ground_truth" /> 	<!-- if not empty, uses tf to retrieve ground truth (has higher priority than gazebo_ground_truth_link) -->
	<arg name="localization_base_link_ground_truth_frame_id" default="base_footprint" />
	<arg name="use_degrees_in_angles" default="false" /> 							<!-- uses radians -->
	<arg name="use_millimeters_in_distances" default="false" /> 					<!-- uses meters -->
	<arg name="use_6_dof" default="true" /> 										<!-- compute error in 3 or 6 dof -->
	<arg name="publish_rate" default="100.0" />										<!-- if > 0 uses regular tf sampling (and also computes error when a new pose arrives), otherwise only computes error when receives poses from topics -->
	<arg name="pose_publishers_sampling_rate" default="10" />						<!-- collect 1 out of pose_publishers_sampling_rate poses (< 0 or 1 -> collects all poses) -->
	<arg name="pose_stamped_topic" default="" />									<!-- empty to disable listening to geometry_msgs::PoseStamped messages -->
	<arg name="pose_stamped_with_covariance_pose_topic" default="/initialpose" />	<!-- empty to disable listening to geometry_msgs::PoseWithCovarianceStamped messages -->
	<arg name="pose_error_publish_topic" default="localization_error" />
	<arg name="tf_lookup_timeout" default="0.15" />
	<arg name="use_time_0_when_querying_tf" default="false" />
	<arg name="localization_poses_output_filename" default="" />
	<arg name="odom_only_poses_output_filename" default="" />
	<arg name="ground_truth_poses_output_filename" default="" />
	<arg name="save_poses_timestamp" default="true" />
	<arg name="save_poses_orientation_quaternion" default="true" />
	<arg name="save_poses_orientation_vector" default="true" />

	<node pkg="robot_localization_tools" type="rlt_robot_localization_error_node" name="rlt_robot_localization_error" output="screen">
		<param name="invert_tf_from_map_ground_truth_frame_id" type="bool" value="false" />
		<param name="map_ground_truth_frame_id" type="str" value="$(arg localization_ground_truth_frame_id)" />
		<param name="map_frame_id" type="str" value="$(arg map_frame_id)" />
		<param name="map_odom_only_frame_id" type="str" value="$(arg map_odom_only_frame_id)" />
		<param name="odom_frame_id" type="str" value="$(arg odom_frame_id)" />
		<param name="base_link_frame_id" type="str" value="$(arg base_link_frame_id)" />
		<param name="base_link_ground_truth_frame_id" type="str" value="$(arg localization_base_link_ground_truth_frame_id)" />
		<param name="publish_rate" type="double" value="$(arg publish_rate)" />
		<param name="pose_publishers_sampling_rate" type="int" value="$(arg pose_publishers_sampling_rate)" />
		<param name="tf_lookup_timeout" type="double" value="$(arg tf_lookup_timeout)" />
		<param name="use_time_0_when_querying_tf" value="$(arg use_time_0_when_querying_tf)" />
		<param name="localization_poses_output_filename" type="str" value="$(arg localization_poses_output_filename)" />
		<param name="odom_only_poses_output_filename" type="str" value="$(arg odom_only_poses_output_filename)" />
		<param name="ground_truth_poses_output_filename" type="str" value="$(arg ground_truth_poses_output_filename)" />
		<param name="save_poses_timestamp" type="bool" value="$(arg save_poses_timestamp)" />
		<param name="save_poses_orientation_quaternion" type="bool" value="$(arg save_poses_orientation_quaternion)" />
		<param name="save_poses_orientation_vector" type="bool" value="$(arg save_poses_orientation_vector)" />
		<param name="pose_stamped_topic" type="str" value="$(arg pose_stamped_topic)" />
		<param name="pose_stamped_with_covariance_pose_topic" type="str" value="$(arg pose_stamped_with_covariance_pose_topic)" />
		<param name="pose_error_publish_topic" type="str" value="$(arg pose_error_publish_topic)" />
		<param name="pose_error_odometry_publish_topic" type="str" value="odometry_error" />
		<param name="localization_poses_publisher_topic" type="str" value="localization_poses" />
		<param name="odometry_poses_publisher_topic" type="str" value="odometry_poses" />
		<param name="ground_truth_poses_publisher_topic" type="str" value="groundtruth_poses" />
		<param name="use_degrees_in_angles" type="bool" value="$(arg use_degrees_in_angles)" />
		<param name="use_millimeters_in_distances" type="bool" value="$(arg use_millimeters_in_distances)" />
		<param name="use_6_dof" type="bool" value="$(arg use_6_dof)" />
	</node>
</launch>
